### PR TITLE
Fix transliterator fallback loading

### PR DESCRIPTION
**Pull request by sffc**
I don't know exactly what the code was doing, but pulling it out into pieces instead of chaining function calls makes it easier to reason about and debug.

### CODE REVIEW

**Review by Manishearth — APPROVED** [review]

File: components/experimental/src/transliterate/transliterator/mod.rs [hunk 1]
```diff
@@ -306,21 +306,31 @@ impl Transliterator {
...
+                let internal_t = match internal_t {
+                    Some(t) => t,
+                    None => {
                         let rbt = Transliterator::load_rbt(
                             #[allow(clippy::unwrap_used)] // infallible
                             DataMarkerAttributes::try_from_str(&dep.to_ascii_lowercase()).unwrap(),
                             transliterator_provider,
                         )?;
-                        Ok(InternalTransliterator::RuleBased(rbt))
-                    })?;
```
Comment (robertbastian) [inline components/experimental/src/transliterate/transliterator/mod.rs:323]: I find this much more readable than the mutable `Option` you're using.

File: components/experimental/src/transliterate/transliterator/mod.rs [hunk 2]
```diff
@@ -341,7 +351,7 @@ impl Transliterator {
     fn load_special<P>(
         special: &str,
         normalizer_provider: &P,
-    ) -> Result<InternalTransliterator, DataError>
+    ) -> Result<Option<InternalTransliterator>, DataError>
```
Comment (robertbastian) [inline components/experimental/src/transliterate/transliterator/mod.rs:354]: I think this should be `Option<Result<InternalTransliterator, DataError>>`. The option signals that a special transliterator is implemented, and then the result is the result from instantiating that transliterator, i.e. the `_` case should just be `None`.

**Review by robertbastian — CHANGES_REQUESTED** [review]
What exactly is the "fix" part of this PR? Can you provide a test case that it fixes?

**Comment by sffc** [comment]
The "fix" is that `load_special` was, I think, always returning an `Err` and not letting the other two functions have a chance at resolving the translit ID (`load_with_override` and `load_rbt`).

File: components/experimental/src/transliterate/transliterator/mod.rs (hunk 1 above)
Comment (sffc) [reply components/experimental/src/transliterate/transliterator/mod.rs:323]: Problems I had with the previous code:

1. It was concealing the bug
2. The functions involved return a mix of Option and Result and it is hard to keep track of which level you are at in unwrapping the values
3. It's difficult to run the debugger with all the closures. You have to dip in and out of the Rust standard library with all the closures.

But I can convert this back to using the option chain if you prefer, as the code owner.

**Comment by sffc** [comment]
> Can you provide a test case that it fixes?

The test that is fixed with this PR is in https://github.com/unicode-org/icu4x/pull/5469. Which course of action do you prefer:

1. Leave it as-is. It is okay to merge this change separate from the demo exercising it.
2. Close this PR and consider landing this fix to be coupled with landing the demo.
3. Make a new test case and add it to this PR.

**Comment by sffc** [comment]
Well, I thought the test I added in #5483 was going to fail on main, but it seems to pass. And I can't reproduce the test failure in #5469 that I was getting last week. :shrug:

**Comment by sffc** [comment]
Made obsolete by #5489

**Pull request closed without merging**